### DISCUSSION TITLE
70. 클러스터형 인덱스와 논클러스터형 인덱스의 차이점을 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#72] [General]
클러스터형 인덱스와 논클러스터형 인덱스의 차이점을 설명해 주세요.

**Comment by eun-byeol** [comment]
클러스터형 인덱스(Clustered  Index)는 `테이블 당 1개`만 생성 가능하고, 일반적으로 `PK로 자동 생성`됩니다.(수동 생성 가능) B+ 트리 구조이며, `리프 페이지에 실제 데이터 페이지를 포함`합니다. 데이터 페이지가 클러스터형 인덱스 기준으로 `물리적 정렬`됩니다. 논클러스터형 인덱스보다 `검색 속도가 더 빠르지만`, 데이터 변경(삽입,삭제,수정) 시 정렬을 유지하므로 `쓰기 성능이 저하`될 수 있습니다.

논클러스터형 인덱스(Non Clustered Index)는 `테이블 당 여러 개` 생성 가능한 보조 인덱스입니다. B+ 트리 구조이며, `리프 페이지`에는 `인덱스 키와 데이터 위치를 나타내는 참조값`(클러스터형 인덱스 키 or RID)으로 구성됩니다. 데이터 페이지가 물리적으로 정렬되지 않고, `인덱스 페이지만 정렬`됩니다. 클러스터형 인덱스보다 `데이터 변경 시 더 빠르지만`, 추가적인 참조 작업으로 `검색 속도가 느릴` 수 있습니다. 또한 `별도의 인덱스 페이지 용량`이 필요한 단점이 있습니다.

**Comment by eun-byeol** [comment]
## 클러스터형 인덱스(Clustered  Index)

![image](https://github.com/user-attachments/assets/39b74377-197d-4c95-9b7a-77e50d7947f5)
- 실제 데이터와 무리(cluster) 지어 인덱싱 됨
- 테이블 당 1개만 생성 가능
    - 일반적으로 PK로 자동 생성되지만 수동 생성 가능
- 데이터와 함께 전체 테이블이 물리적으로 정렬됨
- 루트 페이지(리프 페이지 주소), 리프 페이지(데이터 페이지)로 구성

장점
- 논클러스터형 인덱스보다 읽기 성능이 좋음

단점
- 쓰기 성능이 좋지 않음. 리프 페이지가 모두 차있는데 새로운 데이터를 추가할 때 페이지 분할이 일어남

## 논클러스터형 인덱스(Non Clustered Index)

![image](https://github.com/user-attachments/assets/bcd8975e-a95d-4efb-af30-c57b77b12014)
- 보조 인덱스(Secondary Index)로도 불림
- 테이블 당 여러 개 생성 가능
- **페이지 데이터가 물리적으로 정렬되어 있지 않아** 인덱스에 의해 찾아감
    - **데이터 페이지를 건드리지 않고, 별도의 장소에 인덱스 페이지 생성**
- 리프 페이지에 실제 데이터를 갖지 않음
    - 루트 페이지(리프 페이지 주소), 리프 페이지(데이터 위치를 나타내는 참조값)로 구성
        - 참조값  = `데이터 페이지 번호 + #오프셋` or `클러스터형 인덱스 키`

장점
- 쓰기 성능이 좋음. 리프 페이지가 모두 차 있어도 데이터 페이지 분할이 일어나지 않음
        - 논클러스터드 인덱스 리프 노드에서만 페이지 분할이 일어나고, 데이터 페이지는 분할되지 않음
        - 데이터 물리적 순서와 무관하게 저장되기 때문

단점
- 인덱스 페이지를 위한 추가 저장 공간 필요
- 클러스터형 인덱스보다 읽기 성능이 좋지 않음

## 클러스터형 인덱스 + 논클러스터형 인덱스

![image](https://github.com/user-attachments/assets/c5904c70-11ce-4c23-a505-86ce7243c464)
- 논클러스터형 인덱스 리프 노드에 클러스터형 인덱스 키를 가짐
- 논클러스터형 인덱스를 먼저 거치고, 클러스터형 인덱스를 거쳐 데이터를 찾음

## 혼합형 인덱스 구조에서 논클러스터형 인덱스가 PK값을 가지는 이유

재정렬로 인한 오버헤드를 줄이기 위함. 데이터 변경이 발생하면 테이블 주소가 변경됨. 논클러스터형 인덱스가 실제 저장된 레코드 주소를 가진다면 이 때마다 논클러스터형 인덱스를 모두 수정해야 함.

## InnoDB에서 클러스터링 인덱스

- MySQL에서는 InnoDB 스토리지 엔진에서만 클러스터링 인덱스를 지원
- PK가 없는 경우 우선순위대로 PK를 대체할 칼럼 선택
  1. PK가 있으면 PK 선택
  2. Unique Index + Not null 중 첫 번째 인덱스 선택
  3. 자동으로 유니크한 값을 갖도록 증가되는 칼럼을 내부적으로 추가 후 선택

- 세컨더리 인덱스(논클러스터형 인덱스) 리프 노드가 PK를 가짐
  - MyISAM은 세컨더리 인덱스가 물리적 주소를 가짐→ 리프 노드에서 얻은 주소로 바로 레코드 찾음
  - InnoDB는 논리적 주소(PK)를 가짐→ 리프 노드에서 얻은 주소로 PK 인덱스 검색 후 레코드 찾음
